## Segmentation and Clustering of Toronto Neighborhoods

### Part 1: Using BeautifulSoup to webscrape the Wikipedia article "List of postal codes of Canada"

First, I will import necessary libraries.

In [1]:
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import numpy as np
import pandas as pd
from urllib.request import urlopen

I will then set the article's address equal the the variable url.

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

Next, I use the function `urlopen()` from urllib and set equal to the variable html.

In [3]:
html = urlopen(url)

Using BeautifulSoup, I parse the html and set equal to the variable soup.

In [4]:
soup = BeautifulSoup(html, 'html.parser')

In the article, I right-clicked and chose the "Inspect" element to look at the HTML. There, I saw that the table I need is listed under `<table>` and that the class is listed as `wikitable sortable`.

I use the `soup.find_all()` function to get all tables with class: wikitable sortable from the article. Then, I set that equal to the variable tables.

In [5]:
tables = soup.find_all('table',{'class':'wikitable sortable'})

After doing this, I create three empty lists, one for each column.

Each row is listed under `<tr>`, so I use a for loop using `find_all('tr')` to get all the rows from the table.

Similarly, each cell is listed under `<td>`, so I use a for loop again with `find_all('td')` to get all the cells in each row.

I use `if len(cells) > 1` to skip the first row of the table, which contains the column headers.

Finally, I append the text content of the cells of each column to the corresponding empty list. This can be done using the column's index position. I also strip the text to remove the \n from each cell.

In [6]:
PostalCode = []
Borough = []
Neighborhood = []

for table in tables:
    rows = table.find_all('tr')
    
    for row in rows:
        cells = row.find_all('td')
        
        if len(cells) > 1:
        
            postalcode = cells[0]
            PostalCode.append(postalcode.text.strip())
            
            borough = cells[1]
            Borough.append(borough.text.strip())
            
            neighborhood = cells[2]
            Neighborhood.append(neighborhood.text.strip())

Now I can create the dataframe using `pd.DataFrame`. I include each of the lists, now filled with the cell data from the table, and choose the column names accordingly. 

In [7]:
df = pd.DataFrame(list(zip(PostalCode, Borough, Neighborhood)), columns =['PostalCode', 'Borough', 'Neighborhood']) 

In [8]:
df.head(10)

PostalCode           Borough                                 Neighborhood
0        M1A      Not assigned                                 Not assigned
1        M2A      Not assigned                                 Not assigned
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7        M8A      Not assigned                                 Not assigned
8        M9A         Etobicoke      Islington Avenue, Humber Valley Village
9        M1B       Scarborough                               Malvern, Rouge

I only want the cells with assigned boroughs, so drop the rows where the 'Borough' column has a cell value of 'Not assigned'. I also reset the index here so the rows are properly numbered. 

In [9]:
df1 = df[~df['Borough'].str.contains('Not assigned')].reset_index(drop = True)

I check that the first 10 rows of the dataframe look correct using the `head()` function.

In [10]:
df1.head(10)

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village
6        M1B       Scarborough                               Malvern, Rouge
7        M3B        North York                                    Don Mills
8        M4B         East York              Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto                     Garden District, Ryerson

Finally, I check the shape of the dataframe using the function `shape`. This allows me to see how many rows and columns are contained within df1.

In [11]:
df1.shape

(103, 3)

### Part 2: Getting geospatial data (coordinates) for each postal code

Here, I'll be reading in a CSV of geospatial data for the various neighborhoods.

In [12]:
# The code was removed by Watson Studio for sharing.

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

I want to merge df1 with df_data_1. To do this, I will alter the Postal Code column in df_data_1 so it has the same name as PostalCode in df1, and the merge can be performed on those columns.

In [13]:
df_data_1.rename(columns={'Postal Code': 'PostalCode'}, inplace = True)

Now, I will merge the two dataframes on the PostalCode column using `pd.merge()`. To make sure the dataframe looks like it should after the merge, I will use the `head()` function on the first 10 rows.

In [14]:
merge_df = pd.merge(df1, df_data_1, on='PostalCode')
merge_df.head(10)

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6        M1B       Scarborough                               Malvern, Rouge   
7        M3B        North York                                    Don Mills   
8        M4B         East York              Parkview Hill, Woodbine Gardens   
9        M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

To make sure I have the correct amount of rows after the merge, I will use the `shape` function on merge_df. Remember that there were 103 rows in df1 prior to the merge, so merge_df should have the same amount of rows.

In [15]:
merge_df.shape

(103, 5)

### Part 3: Exploring and clustering the neighborhoods of Toronto

In this section, I will be working with the Toronto neighborhoods to generate clusters.

First, I will use the function `str.contains` to select the rows where the cell value in 'Borough' contains the string 'Toronto'. Once again, I reset the index to probably number the rows.

In [16]:
tor_df = merge_df[merge_df['Borough'].str.contains('Toronto')].reset_index(drop = True)

In [17]:
tor_df

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   
15        M4L      East Toronto   
16        M5L  Downtown Toronto   
17        M4M      East Toronto   
18        M4N   Central Toronto   
19        M5N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
24        M5R   Central Toronto   
25        M6R      West Toronto   
26        M4S   Central Toronto   
27        M5S  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
30        M5T  Downtown Toronto   
31        M4V   Central Toronto   
32        M5V  Downtown Toronto   
33        M4W  Downtown Toronto   
34        M5W  Downtown Toronto   
35        M4X  Downtown Toronto   
36        M5X  Downtown Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                 North Toronto West,  Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
32  CN Tower, King and Spadi

Here, I want to create a map of Toronto, so I'll need to import the necessary libraries.

In [18]:
# To convert an address into latitude and longitude values
from geopy.geocoders import Nominatim 

# To handle requests
import requests

# To transform JSON into a pandas dataframe
from pandas import json_normalize

# Plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# To import k-means from clustering stage
from sklearn.cluster import KMeans

# To render maps
!pip install folium
import folium

print('Libraries imported.')

Libraries imported.


Next, I want to get the coordinates for Toronto. I can use the geopy library to do this.

In [19]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinates of Toronto are 43.6534817, -79.3839347.


Now I'll create a map of Toronto with the neighborhoods superimposed on top using the folium library.

In [20]:
toronto_map = folium.Map(location = [latitude, longitude], zoom_start = 12)

for lat, lng, borough, neighborhood in zip(tor_df['Latitude'], tor_df['Longitude'], tor_df['Borough'], tor_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
    [lat, lng],
    radius = 5,
    popup = label,
    color = 'indigo',
    fill = True,
    fill_color = '#3186cc',
    fill_opacity = 0.7,
    parse_html = False).add_to(toronto_map)
    
toronto_map

Now that I have the map, I'm going to use the Foursquare API to explore and segment the neighborhoods.

In [21]:
# The code was removed by Watson Studio for sharing.

Now, I'll explore the neighborhoods a little. First I'm going to select the first neighborhood in the dataframe and get its coordinates.

In [22]:
neighborhood_name = tor_df.loc[0, 'Neighborhood']
neighborhood_latitude = tor_df.loc[0, 'Latitude']
neighborhood_longitude = tor_df.loc[0, 'Longitude']

print('The latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, neighborhood_latitude, neighborhood_longitude))

The latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


Next, I'll use the Foursquare API to take a look at the venues surround the Regent Park, Harbourfront area. For this analysis I'll limit it to 100 results in a 500 meter radius.

In [23]:
LIMIT = 100
radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=XUKV0ERQ5N5GHGT4IL4ZY4OQXFV1KK5IQV4WAI5CEQMR25JB&client_secret=E5TQFWQGWKD5S4WQOFZI5B1WAP0TXEXBBPXMF1LP51KXAC1W&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

The results will be returned in a JSON format.

In [24]:
results = requests.get(url).json()

Next, I'll create a function that extracts the category of the venue - i.e. what type of business the venue is.

In [25]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

I will clean the JSON data and structure it into a dataframe.

In [26]:
venues = results['response']['groups'][0]['items']

nearby_venues = json_normalize(venues)

filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis = 1)

nearby_venues.columns = [col.split('.')[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [27]:
print('Foursquare returned {} venues.'.format(nearby_venues.shape[0]))

Foursquare returned 44 venues.


Now, I'll create a function that allows me to repeat this for all the neighborhoods in Toronto.

In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    venues_list = []
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
    
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        venues_list.append([(
            name,
            lat,
            lng,
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
    
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
    
    return(nearby_venues)

In [29]:
toronto_venues = getNearbyVenues(names = tor_df['Neighborhood'], latitudes = tor_df['Latitude'], longitudes = tor_df['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West,  Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport


We can check how many venues were returned for each neighborhood...

In [30]:
toronto_venues.groupby(['Neighborhood']).count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West,  Lawrence Park                                     18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         96   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

...and how many unique categories there are.

In [31]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 236 uniques categories.


Now I'll start encoding and analyzing the neighborhoods in order to cluster them.

In [32]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theater  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...        0                 0                 0      0   
1             0  ...        0                 0                 0      0   
2             0  ...        0                 0                 0      0   
3             0  ...        0                 0                 0      0   
4             0  ...        0                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  

[5 rows x 236 columns]

I group the rows by neighborhood and take the mean of the frequency of occurence of each category.

In [33]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.014706   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.026667   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022222   
20                        Moore Park, Summerhill East     0.000000   
21                 North Toronto West,  Lawrence Park     0.055556   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022727   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.030303   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010417   
32                                    Studio District     0.027027   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.029412   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0625              0.0625        0.0625   
4            0.000000   0.0000              0.0000        0.0000   
5            0.000000   0.0000              0.0000        0.0000   
6            0.013333   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.000000   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.0

Next, I'll print each neighborhood and its top 5 most common venues.

In [34]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1  Seafood Restaurant  0.04
2         Cheese Shop  0.04
3            Beer Bar  0.04
4        Cocktail Bar  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1       Nightclub  0.09
2     Coffee Shop  0.09
3  Breakfast Spot  0.09
4   Grocery Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0  Fast Food Restaurant  0.06
1         Auto Workshop  0.06
2            Comic Shop  0.06
3           Pizza Place  0.06
4      Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0   Airport Lounge  0.12
1  Airport Service  0.12
2  Harbor / Marina  0.06
3            Plane  0.06
4    Boat or Ferry  0.06


----Central Bay Street----
                venue  freq
0         Coff

I will turn this into a pandas dataframe by first writing a function to sort the venues in descending order, and then creating the dataframe to display the top 10 venues for each neighborhood.

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...            Skate Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market                Bakery          Cocktail Bar   
1           Coffee Shop        Breakfast Spot             Nightclub   
2           Pizza Place               Brewery         Burrito Place   
3       Airport Service           Coffee Shop       Harbor / Marina   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant           Cheese Shop              Beer Bar   
1             Pet Store                   Bar         Burrito Place   
2            Restaurant        Farmers Market  Fast Food Restaurant   
3                 Plane   Rental Car Location      Sculpture Garden   
4           Salad Place          Burger Joint      Department Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant         Grocery Store                    Pub  
1            Restaurant          Climbing Gym  Performing Arts Venue  
2      Recording Studio               Butcher     Light Rail Station  
3              Boutique                   Bar          Boat or Ferry  
4   Japanese Restaurant       Thai Restaurant        Bubble Tea Shop

Now that the results in a dataframe, I can begin to cluster the neighborhoods using k-means.

In [37]:
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

kmeans = KMeans(n_clusters = kclusters, random_state = 0).fit(toronto_grouped_clustering)

kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Now I'll create a new dataframe that includes both the cluster labels and the top 10 venues.

In [38]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = tor_df

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on = 'Neighborhood')

toronto_merged.head()

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0           Coffee Shop   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               0     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                  Park                Bakery   
1           Yoga Studio            Hobby Shop            Restaurant   
2        Clothing Store                  Café   Japanese Restaurant   
3                  Café          Cocktail Bar            Restaurant   
4                   Pub                 Trail               Dog Run   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0                  Café               Theater             Breakfast Spot   
1        Sandwich Place   Distribution Center                      Diner   
2       Bubble Tea Shop        Cosmetics Shop  Middle Eastern Restaurant   
3             Gastropub   American Restaurant                   Beer Bar   
4          Dessert Shop    Dim Sum Restaurant                      Diner   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Brewery                   Spa             Beer Store  
1      Italian Restaurant              Beer Bar                   Café  
2  Furniture / Home Store        Lingerie Store            Pizza Place  
3          Clothing Store                   Gym         Cosmetics Shop  
4          Discount Store   Distribution Center          Women's Store

Now, we can visualize the clusters on the map.

In [39]:
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_hmtl = True)
    folium.CircleMarker(
    [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Each cluster can be viewed in an individual dataframe

#### Cluster 1:

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
0   Downtown Toronto               0                    Coffee Shop   
1   Downtown Toronto               0                    Coffee Shop   
2   Downtown Toronto               0                    Coffee Shop   
3   Downtown Toronto               0                    Coffee Shop   
4       East Toronto               0              Health Food Store   
5   Downtown Toronto               0                    Coffee Shop   
6   Downtown Toronto               0                    Coffee Shop   
7   Downtown Toronto               0                  Grocery Store   
8   Downtown Toronto               0                    Coffee Shop   
9       West Toronto               0                       Pharmacy   
10  Downtown Toronto               0                    Coffee Shop   
11      West Toronto               0                            Bar   
12      East Toronto               0               Greek Restaurant   
13  Downtown Toronto               0                    Coffee Shop   
14      West Toronto               0                           Café   
15      East Toronto               0                           Park   
16  Downtown Toronto               0                    Coffee Shop   
17      East Toronto               0                    Coffee Shop   
20   Central Toronto               0                           Park   
22      West Toronto               0             Mexican Restaurant   
23   Central Toronto               0                    Coffee Shop   
24   Central Toronto               0                 Sandwich Place   
25      West Toronto               0                      Gift Shop   
26   Central Toronto               0                    Pizza Place   
27  Downtown Toronto               0                           Café   
28      West Toronto               0                           Café   
30  Downtown Toronto               0  Vegetarian / Vegan Restaurant   
31   Central Toronto               0                    Coffee Shop   
32  Downtown Toronto               0                 Airport Lounge   
34  Downtown Toronto               0                    Coffee Shop   
35  Downtown Toronto               0                    Coffee Shop   
36  Downtown Toronto               0                    Coffee Shop   
37  Downtown Toronto               0                    Coffee Shop   
38      East Toronto               0                     Skate Park   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                   Park                Bakery   
1            Yoga Studio             Hobby Shop            Restaurant   
2         Clothing Store                   Café   Japanese Restaurant   
3                   Café           Cocktail Bar            Restaurant   
4                    Pub                  Trail               Dog Run   
5         Farmers Market                 Bakery          Cocktail Bar   
6                   Café         Sandwich Place    Italian Restaurant   
7                   Café                   Park           Candy Store   
8                   Café                  Hotel            Restaurant   
9                 Bakery          Grocery Store               Brewery   
10              Aquarium                   Café                 Hotel   
11           Coffee Shop  Vietnamese Restaurant            Restaurant   
12           Coffee Shop     Italian Restaurant            Restaurant   
13                 Hotel             Restaurant                  Café   
14           Coffee Shop         Breakfast Spot             Nightclub   
15        Sandwich Place   Fast Food Restaurant     Fish & Chips Shop   
16            Restaurant                   Café                 Hotel   
17                Bakery              Gastropub               Brewery   
20                 Hotel         Breakfast Spot        Sandwich Place   
22                  Café                    Bar       Thai Restaura

#### Cluster 2:

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               1                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29            Playground         Women's Store          Dance Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 3:

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18              Bus Line           Swim School    College Rec Center   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
18      Department Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18            Donut Shop      Doner Restaurant                Dog Run

#### Cluster 4:

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21   Central Toronto               3                  Park   
33  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21      Sushi Restaurant         Jewelry Store                 Trail   
33                 Trail            Playground          Dance Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
33  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21      Doner Restaurant               Dog Run    Distribution Center  
33      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 5:

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4           Music Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden         Women's Store         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
19     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19            Donut Shop      Doner Restaurant                Dog Run